File: Demos/Kotlin Demos/13. ShakeDetector/app/src/main/java/org/feup/apm/shakedetector/MainActivity.java
package org.feup.apm.shakedetector;

import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity implements ShakeDetector.OnShakeDetectedListener {
  LinearLayout root;
  TextView tvSensorName;
  boolean shake = false;
  ShakeDetector shakeDetector;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    root = (LinearLayout) findViewById(R.id.root);
    root.setBackgroundResource(R.color.colorShake1);
    tvSensorName = (TextView) findViewById(R.id.tvSName);
    shakeDetector = new ShakeDetector(this);
    shakeDetector.setOnShakeDetectedListener(this);
  }

  @Override
  protected void onResume() {
    super.onResume();
    String sName = shakeDetector.startSensing();
    if (sName != null)
      tvSensorName.setText("("+sName+")");
    else {
      Toast.makeText(this, "There is no linear accel sensor!", Toast.LENGTH_LONG).show();
      finish();
    }
  }

  @Override
  protected void onPause() {
    super.onPause();
    shakeDetector.stopSensing();
  }

  public void toggleBackground() {
    if (shake)
      root.setBackgroundResource(R.color.colorShake1);
    else
      root.setBackgroundResource(R.color.colorShake2);
    shake = !shake;
  }

  public void onShakeDetected() {
    toggleBackground();
  }
}

File: Demos/NFCDemos/NFCReader/app/src/main/java/org/feup/apm/nfcreader/MainActivity.kt
package org.feup.apm.nfcreader

import android.nfc.NfcAdapter
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity

const val READER_FLAGS = NfcAdapter.FLAG_READER_NFC_A or NfcAdapter.FLAG_READER_SKIP_NDEF_CHECK

class MainActivity : AppCompatActivity(), CardReader.CardReaderCallback {
  val nfc by lazy { NfcAdapter.getDefaultAdapter(this) }
  val tvAccNr by lazy { findViewById<TextView>(R.id.card_account_field) }
  val cardReader by lazy { CardReader(this) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    if (nfc == null) {
      Toast.makeText(this, "NFC adapter not present!", Toast.LENGTH_LONG).show()
      finish()
    }
    findViewById<Button>(R.id.bt_clear).setOnClickListener { tvAccNr.setText(R.string.tv_accnr) }
  }

  override fun onPause() {
    super.onPause()
    nfc.disableReaderMode(this)
  }

  override fun onResume() {
    super.onResume()
    nfc.enableReaderMode(this, cardReader, READER_FLAGS, null)
  }

  override fun onCardNrReceived(accountNr: String) {
    runOnUiThread { tvAccNr.setText(accountNr) }
  }
}